File: library/src/main/java/org/xmtp/android/library/codecs/RemoteAttachmentCodec.kt
package org.xmtp.android.library.codecs

import com.google.protobuf.ByteString
import com.google.protobuf.kotlin.toByteString
import com.google.protobuf.kotlin.toByteStringUtf8
import org.web3j.crypto.Hash
import org.web3j.utils.Numeric
import org.xmtp.android.library.Crypto
import org.xmtp.android.library.XMTPException
import org.xmtp.android.library.toHex
import org.xmtp.proto.message.contents.CiphertextOuterClass.Ciphertext
import java.net.URI
import java.net.URL
import java.security.SecureRandom

data class EncryptedEncodedContent(
    val contentDigest: String,
    val secret: ByteString,
    val salt: ByteString,
    val nonce: ByteString,
    val payload: ByteString,
    val contentLength: Int? = null,
    val filename: String? = null,
)

data class RemoteAttachment(
    val url: URL,
    val contentDigest: String,
    val secret: ByteString,
    val salt: ByteString,
    val nonce: ByteString,
    val scheme: String,
    var contentLength: Int? = null,
    var filename: String? = null,
    var fetcher: Fetcher = HTTPFetcher(),
) {
    fun <T> load(): T? {
        val payload = fetcher.fetch(url)

        if (payload.isEmpty()) {
            throw XMTPException("no remote attachment payload")
        }

        val encrypted = EncryptedEncodedContent(
            contentDigest,
            secret,
            salt,
            nonce,
            payload.toByteString(),
            contentLength,
            filename,
        )

        val decrypted = decryptEncoded(encrypted)

        return decrypted.decoded<T>()
    }

    companion object {
        fun decryptEncoded(encrypted: EncryptedEncodedContent): EncodedContent {
            if (Hash.sha256(encrypted.payload.toByteArray()).toHex() != encrypted.contentDigest) {
                throw XMTPException("contentDigest does not match")
            }

            val aes = Ciphertext.Aes256gcmHkdfsha256.newBuilder().also {
                it.hkdfSalt = encrypted.salt
                it.gcmNonce = encrypted.nonce
                it.payload = encrypted.payload
            }.build()

            val ciphertext = Ciphertext.newBuilder().also {
                it.aes256GcmHkdfSha256 = aes
            }.build()

            val decrypted = Crypto.decrypt(encrypted.secret.toByteArray(), ciphertext)

            return EncodedContent.parseFrom(decrypted)
        }

        fun <T> encodeEncrypted(content: T, codec: ContentCodec<T>): EncryptedEncodedContent {
            val secret = SecureRandom().generateSeed(32)
            val encodedContent = codec.encode(content).toByteArray()
            val ciphertext = Crypto.encrypt(secret, encodedContent) ?: throw XMTPException("ciphertext not created")
            val contentDigest = Hash.sha256(ciphertext.aes256GcmHkdfSha256.payload.toByteArray()).toHex()
            return EncryptedEncodedContent(
                contentDigest = contentDigest,
                secret = secret.toByteString(),
                salt = ciphertext.aes256GcmHkdfSha256.hkdfSalt,
                nonce = ciphertext.aes256GcmHkdfSha256.gcmNonce,
                payload = ciphertext.aes256GcmHkdfSha256.payload,
                contentLength = null,
                filename = null,
            )
        }

        fun from(url: URL, encryptedEncodedContent: EncryptedEncodedContent): RemoteAttachment {
            if (URI(url.toString()).scheme != "https") {
                throw XMTPException("scheme must be https://")
            }

            return RemoteAttachment(
                url = url,
                contentDigest = encryptedEncodedContent.contentDigest,
                secret = encryptedEncodedContent.secret,
                salt = encryptedEncodedContent.salt,
                nonce = encryptedEncodedContent.nonce,
                scheme = URI(url.toString()).scheme,
            )
        }
    }
}

val ContentTypeRemoteAttachment = ContentTypeIdBuilder.builderFromAuthorityId(
    "xmtp.org",
    "remoteStaticAttachment",
    versionMajor = 1,
    versionMinor = 0
)

interface Fetcher {
    fun fetch(url: URL): ByteArray
}

class HTTPFetcher : Fetcher {
    override fun fetch(url: URL): ByteArray {
        return url.readBytes()
    }
}

data class RemoteAttachmentCodec(override var contentType: ContentTypeId = ContentTypeRemoteAttachment) : ContentCodec<RemoteAttachment> {
    override fun encode(content: RemoteAttachment): EncodedContent {
        return EncodedContent.newBuilder().also {
            it.type = ContentTypeRemoteAttachment
            it.putAllParameters(
                mapOf(
                    "contentDigest" to content.contentDigest,
                    "secret" to content.secret.toByteArray().toHex(),
                    "salt" to content.salt.toByteArray().toHex(),
                    "nonce" to content.nonce.toByteArray().toHex(),
                    "scheme" to content.scheme,
                    "contentLength" to content.contentLength.toString(),
                    "filename" to content.filename,
                )
            )
            it.content = content.url.toString().toByteStringUtf8()
        }.build()
    }

    override fun decode(content: EncodedContent): RemoteAttachment {
        val contentDigest = content.parametersMap["contentDigest"] ?: throw XMTPException("missing content digest")
        val secret = content.parametersMap["secret"] ?: throw XMTPException("missing secret")
        val salt = content.parametersMap["salt"] ?: throw XMTPException("missing salt")
        val nonce = content.parametersMap["nonce"] ?: throw XMTPException("missing nonce")
        val scheme = content.parametersMap["scheme"] ?: throw XMTPException("missing scheme")
        val contentLength = content.parametersMap["contentLength"] ?: throw XMTPException("missing contentLength")
        val filename = content.parametersMap["filename"] ?: throw XMTPException("missing filename")
        val encodedContent = content.content ?: throw XMTPException("missing content")

        return RemoteAttachment(
            url = URL(encodedContent.toStringUtf8()),
            contentDigest = contentDigest,
            secret = Numeric.hexStringToByteArray(secret).toByteString(),
            salt = Numeric.hexStringToByteArray(salt).toByteString(),
            nonce = Numeric.hexStringToByteArray(nonce).toByteString(),
            scheme = scheme,
            contentLength = contentLength.toInt(),
            filename = filename,
        )
    }
}

File: Sources/XMTP/Codecs/RemoteAttachmentCodec.swift
//
//  RemoteAttachmentCodec.swift
//
//
//  Created by Pat Nakajima on 2/19/23.
//

import CryptoKit
import Foundation
import web3

public let ContentTypeRemoteAttachment = ContentTypeID(authorityID: "xmtp.org", typeID: "remoteStaticAttachment", versionMajor: 1, versionMinor: 0)

public enum RemoteAttachmentError: Error {
	case invalidURL, v1NotSupported, invalidParameters(String), invalidDigest(String), invalidScheme(String), payloadNotFound
}

protocol RemoteContentFetcher {
	func fetch(_ url: String) async throws -> Data
}

struct HTTPFetcher: RemoteContentFetcher {
	func fetch(_ url: String) async throws -> Data {
		guard let url = URL(string: url) else {
			throw RemoteAttachmentError.invalidURL
		}

		return try await URLSession.shared.data(from: url).0
	}
}

public struct RemoteAttachment {
	public enum Scheme: String {
		case https = "https"
	}

	public var url: String
	public var contentDigest: String
	public var secret: Data
	public var salt: Data
	public var nonce: Data
	public var scheme: Scheme
	var fetcher: RemoteContentFetcher

	// Non standard
	public var contentLength: Int?
	public var filename: String?

	public init(url: String, contentDigest: String, secret: Data, salt: Data, nonce: Data, scheme: Scheme) throws {
		self.url = url
		self.contentDigest = contentDigest
		self.secret = secret
		self.salt = salt
		self.nonce = nonce

		self.scheme = scheme
		fetcher = HTTPFetcher()

		try ensureSchemeMatches()
	}

	public init(url: String, encryptedEncodedContent: EncryptedEncodedContent) throws {
		self.url = url
		contentDigest = encryptedEncodedContent.digest
		secret = encryptedEncodedContent.secret
		salt = encryptedEncodedContent.salt
		nonce = encryptedEncodedContent.nonce

		scheme = .https
		fetcher = HTTPFetcher()

		try ensureSchemeMatches()
	}

	func ensureSchemeMatches() throws {
		if !url.hasPrefix(scheme.rawValue) {
			throw RemoteAttachmentError.invalidScheme("scheme must be https")
		}
	}

	public static func encodeEncrypted<Codec: ContentCodec, T>(content: T, codec: Codec, with client: Client) throws -> EncryptedEncodedContent where Codec.T == T {
		let secret = try Crypto.secureRandomBytes(count: 32)
		let encodedContent = try codec.encode(content: content, client: client).serializedData()
		let ciphertext = try Crypto.encrypt(secret, encodedContent)
		let contentDigest = sha256(data: ciphertext.aes256GcmHkdfSha256.payload)

		return EncryptedEncodedContent(
			secret: secret,
			digest: contentDigest,
			salt: ciphertext.aes256GcmHkdfSha256.hkdfSalt,
			nonce: ciphertext.aes256GcmHkdfSha256.gcmNonce,
			payload: ciphertext.aes256GcmHkdfSha256.payload
		)
	}

	public static func decryptEncoded(encrypted: EncryptedEncodedContent) throws -> EncodedContent {
		if RemoteAttachment.sha256(data: encrypted.payload) != encrypted.digest {
			throw RemoteAttachmentError.invalidDigest("content digest does not match. expected: \(encrypted.digest), got: \(SHA256.hash(data: encrypted.payload).description)")
		}

		let decrypted = try Crypto.decrypt(encrypted.secret, CipherText.with {
			let aes256GcmHkdfSha256 = CipherText.Aes256gcmHkdfsha256.with { aes in
				aes.hkdfSalt = encrypted.salt
				aes.gcmNonce = encrypted.nonce
				aes.payload = encrypted.payload
			}

			$0.aes256GcmHkdfSha256 = aes256GcmHkdfSha256
		})
		return try EncodedContent(serializedData: decrypted)
	}

	public func content() async throws -> EncodedContent {
		let payload = try await fetcher.fetch(url)

		if payload.isEmpty {
			throw RemoteAttachmentError.payloadNotFound
		}

		return try RemoteAttachment.decryptEncoded(encrypted: EncryptedEncodedContent(
				secret: secret,
				digest: contentDigest,
				salt: salt,
				nonce: nonce,
				payload: payload
		))
	}

	private static func sha256(data: Data) -> String {
		Data(SHA256.hash(data: data)).toHex
	}
}

public struct RemoteAttachmentCodec: ContentCodec {
    
	public typealias T = RemoteAttachment

	public init() {}

	public var contentType = ContentTypeRemoteAttachment

	public func encode(content: RemoteAttachment, client _: Client) throws -> EncodedContent {
		var encodedContent = EncodedContent()

		encodedContent.type = ContentTypeRemoteAttachment
		encodedContent.content = Data(content.url.utf8)
		encodedContent.parameters = [
			"contentDigest": content.contentDigest,
			"secret": content.secret.toHex,
			"salt": content.salt.toHex,
			"nonce": content.nonce.toHex,
			"scheme": "https",
			"contentLength": String(content.contentLength ?? -1),
			"filename": content.filename ?? "",
		]

		return encodedContent
	}

	public func decode(content: EncodedContent, client _: Client) throws -> RemoteAttachment {
		guard let url = String(data: content.content, encoding: .utf8) else {
			throw RemoteAttachmentError.invalidURL
		}

		guard let contentDigest = content.parameters["contentDigest"] else {
			throw RemoteAttachmentError.invalidDigest("missing contentDigest parameter")
		}

		let secret = try getHexParameter("secret", from: content.parameters)
		let salt = try getHexParameter("salt", from: content.parameters)
		let nonce = try getHexParameter("nonce", from: content.parameters)

		guard let schemeString = content.parameters["scheme"] else {
			throw RemoteAttachmentError.invalidScheme("no scheme parameter")
		}

        if (!schemeString.hasPrefix(RemoteAttachment.Scheme.https.rawValue)) {
			throw RemoteAttachmentError.invalidScheme("invalid scheme value. must start with https")
		}

		var attachment = try RemoteAttachment(url: url, contentDigest: contentDigest, secret: secret, salt: salt, nonce: nonce, scheme: RemoteAttachment.Scheme.https)

		if let contentLength = content.parameters["contentLength"] {
			attachment.contentLength = Int(contentLength)
		}

		if let filename = content.parameters["filename"] {
			attachment.filename = filename
		}

		return attachment
	}
    
    public func fallback(content: RemoteAttachment) throws -> String? {
        return "Can’t display “\(String(describing: content.filename))”. This app doesn’t support attachments."
    }

	private func getHexParameter(_ name: String, from parameters: [String: String]) throws -> Data {
		guard let parameterHex = parameters[name] else {
			throw RemoteAttachmentError.invalidParameters("missing \(name) parameter")
		}

		guard let parameterData = parameterHex.web3.hexData else {
			throw RemoteAttachmentError.invalidParameters("invalid \(name) value")
		}

		return Data(parameterData)
	}
}
